proposals/p2875.md

```diff
@@ -0,0 +1,475 @@
+# Functions, function types, and function calls
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2875)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Functions and function types](#functions-and-function-types)
+    -   [Call syntax](#call-syntax)
+    -   [Direct calls](#direct-calls)
+    -   [Generic callable parameters](#generic-callable-parameters)
+    -   [Function types and `Call`](#function-types-and-call)
+    -   [Overloaded call operator](#overloaded-call-operator)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Signature-based function types](#signature-based-function-types)
+    -   [Make direct and indirect calls behave uniformly](#make-direct-and-indirect-calls-behave-uniformly)
+-   [Future work](#future-work)
+    -   [Overloading](#overloading)
+    -   [Lambdas](#lambdas)
+    -   [Function pointers](#function-pointers)
+
+<!-- tocstop -->
+
+## Abstract
+
+Specify the behavior of function calls and the type and behavior of the entity
+introduced by a function declaration.
+
+## Problem
+
+Functions are a primitive building block through which we define Carbon's
+execution semantics, but we don't currently have a specification for what a
+function is in Carbon nor how function calls work. We have open questions around
+whether Carbon has first-class function types, how types can overload the
+function call operation, and how callable entities such as the name of a
+parameterized class fit into the picture -- are they functions or something
+else?
+
+## Background
+
+This proposal assumes the reader understands how functions and function pointers
+work in C or C++.
+
+## Proposal
+
+A function declaration in Carbon introduces a new, unique type, called a
+_function type_. The function name is a stateless value of that function type.
```

Comment:
In some ways this compromises our "types are values" philosophy, because we're introducing both a type and a singular value of that type, in order to represent the same information in the type and value domains. I don't see how we can avoid that, though, so this is just an observation; no action is needed.